langserver/completion.go

```diff
@@ -155,41 +158,50 @@ func (s *server) completeFunctionName(completions *[]protocol.CompletionItem, lo
 		return err
 	}
 
+	names := make([]string, len(promql.Functions))
+	i := 0
 	for name := range promql.Functions {
-		if strings.HasPrefix(strings.ToLower(name), metricName) {
-			item := protocol.CompletionItem{
-				Label:            name,
-				SortText:         "__1__" + name,
-				Kind:             3, //Function
-				InsertTextFormat: 2, //Snippet
-				TextEdit: &protocol.TextEdit{
-					Range:   editRange,
-					NewText: name + "($1)",
-				},
-				Command: &protocol.Command{
-					// This might create problems with non VS Code clients
-					Command: "editor.action.triggerParameterHints",
-				},
-			}
-			*completions = append(*completions, item)
+		names[i] = strings.ToLower(name)
+		i++
+	}
+
+	for _, match := range fuzzy.Find(metricName, names) {
```

Comment:
since this loop is repeated three times in this file, you could create a function in order to reuse this loop and this logic

Reply:
sounds good to me. i will see to it.

Reply:
so @Nexucis i tried factorising this loop to a single function but it get complicated  exponentially . Reason is mostly due to the dynamic map type of each the completion type for instance the map type for the metrics names are `map[string][]v1.Metadata` while for functions names are  `map[string]*Function{...}` etc. and since the map is required to get a description it is necessary to pass the map.
That said the map can be passed as an empty interface{} but a type switch would be required to get the underling map value hence duplicating the loop again in the function.

Reply:
It might be interesting to look at how gopls is approaching this and the sorting issue:

https://github.com/golang/tools/blob/52effbd89c51c477f75e877aa5ded301b29fe20d/internal/lsp/completion.go#L71-L124

Reply:
for the sorting the trick they use is pretty neat [this line](https://github.com/golang/tools/blob/52effbd89c51c477f75e877aa5ded301b29fe20d/internal/lsp/completion.go#L112) it prepends zeros the score. i guess we could use that approach ?

Reply:
It's a bit hacky, since they never know for sure how many zeros are needed. But it could be acceptable.

A correct implementation would first need to check how many leading zeros are required at max.

Reply:
I think maximum length of the strings in the map keys could tell how many leading zeros are needed then?

Also this type [CompletionItem](https://github.com/golang/tools/blob/52effbd89c51c477f75e877aa5ded301b29fe20d/internal/lsp/source/completion.go#L31) could help in proper factoring

Reply:
Idk, if it's a good idea to introduce a second `CompletionItem` type that looks almost as the first one. What would be the benefit?

Reply:
factoring as suggested by @Nexucis but not without its quirks. i think the current implementation keep things simple enough

Reply:
Sounds reasonable to me.